### ISSUE TITLE
Can not merge tablets with write ahead logs

### ISSUE THREAD

**Issue by keith-turner** [apache/accumulo#3845]
Tablets with write ahead logs should not be merged, see #3844 for details on why.  If a tablet has write ahead logs, then merge should wait for those to be processed before proceeding.  Need to see if the existing merge code defends against this, if this may be a problem for more than the elasticity branch.

**Comment by keith-turner** [comment]
@cshannon do you recall if the existing merge code looked for write ahead logs and had and special handling for them?

**Comment by cshannon** [comment]
@keith-turner - I don't remember seeing anything related to write ahead logs in the existing merge code and I was in there quite a bit so it may not be specifically handled.  I can take a look tomorrow though and double check and see if there is something that would prevent merge but I don't think there is.

**Comment by keith-turner** [comment]
>  I can take a look tomorrow though and double check and see if there is something that would prevent merge but I don't think there is.

If there is not, then may need to  set this as a 2.1.x bug.

**Comment by cshannon** [comment]
I looked into this some today and the merge code will first make sure that the tablets are hosted before beginning the steps to start the merge to try and prevent any wals from existing as the wals will be processed before the tablet is hosted. After the merge goes into the `started` state it verifies the total tablets equals hosted before going into the waiting for offline state. https://github.com/apache/accumulo/blob/e68fa4ac2243094c64451e06c554653706a37b64/server/manager/src/main/java/org/apache/accumulo/manager/state/MergeStats.java#L101-L110

After talking to @keith-turner , it was pointed out that there is a potential edge case where during the period of time that the manager is waiting for the tablets to be offline to merge (so after the hosted check) wals could exist that need recovery if a tserver dies. To handle the edge case the merge could check to make sure that all the unhosted tablets do not have any wals before beginning the metadata changes and if wals exist throw an exception to abort the merge.

I can work on a fix for this which would need to also go to 2.1 and main and not just elasticity

**Comment by keith-turner** [comment]
@cshannon  I looked around in the code and suspect the following strategy may work to fix this.

 * MergeStats tracks the number of wals seen
 * MergeStats.nextMergeState() does not transition out of MergeState.WAITING_FOR_OFFLINE if the wal count >0 for any tablet
 * Manager.getGoalState() will return hosted when a tablet has wals even if merge state is WAITING_FOR_OFFLINE.  May be able to get walog count from the passed in TabletLocationState.
 * When actually merging tablets, if wals are seen then throw an exception as this is not expected (should not get that far with wals).

Was looking at the code in main when making the above comments.  Hoping WAITING_FOR_OFFLINE is common code for both merge and delete rows, did not check that.

**Comment by cshannon** [comment]
I'll take a look later this week and see if that strategy will work and try and come up with a PR if it will.

**Comment by cshannon** [comment]
@keith-turner - I started looking at this and comparing 2.1 and main for what needs to happen for a fix. While looking at 2.1 I realized that the 2.1 branch (with the old chop compactions) does in fact already do the first 3 items on your list but **only on tablets that need to be chopped**, not all tablets as part of the merge. These checks do not exist in main because they were removed with no-chop merge being implemented.

#### Here is where merge stats is tracking that wals exist with a flag in 2.1 (not a count just that they exist) and then preventing the transition in the verification check for tablets needing to be chopped:

https://github.com/apache/accumulo/blob/324f52b711db908f144c943bcf1bc077af73e39b/server/manager/src/main/java/org/apache/accumulo/manager/state/MergeStats.java#L101-L108

https://github.com/apache/accumulo/blob/324f52b711db908f144c943bcf1bc077af73e39b/server/manager/src/main/java/org/apache/accumulo/manager/state/MergeStats.java#L233-L236

#### And here is the getGoalState() method verifying wals in 2.1 are empty before setting the goal to UNASSIGNED when waiting for chopped:

https://github.com/apache/accumulo/blob/324f52b711db908f144c943bcf1bc077af73e39b/server/manager/src/main/java/org/apache/accumulo/manager/Manager.java#L650-L659

I know in #3844 you mentioned that for elasticity things are unique and need to prevent merge/split as you wouldn't be able to find the WALs due to the manager doing the splits. So does that mean that for 2.1 the checks on just for the tablets that need chopping is good enough? This is still new enough to me that I'm not quite sure.

If 2.1 is ok than that means the main fix here is to restore the checks for 3.1 (i guess on all tablets still?) and also to add the verification before the actual merge in both 2.1 and 3.1. Otherwise if you think we should still verify no WALs for all tablets, even in 2.1, I can do that as well pretty easily.

Thoughts?

**Comment by keith-turner** [comment]
@cshannon I have been looking at the 2.1 and 3.1 code.  As an overall goal for 2.1.x and 3.1, do the following goals seem correct?

 1. Only transition from WAITING_FOR_OFFLINE to MERGE if all tablets have no locations and no wals.
 2. Throw an exception if a location or wal is seen in the metadata table when in the MERGE state.

For goal 1, the 2.1 code does seems mostly correct.  However when a tablet has wals and the merge state is WAITING_FOR_OFFLINE I am not sure 2.1 would host that tablet so it could be recovered.  Seems like [this code](https://github.com/apache/accumulo/blob/9dfa9e3b6489a8dc7e00d9377680bb91ef87b242/server/manager/src/main/java/org/apache/accumulo/manager/Manager.java#L660) may keep the tablet unassigned when it has wals which would prevent recovery so maybe in 2.1 a merge could get stuck. For 3.1, it does seems that modifications are needed for goal 1 as you mentioned.

For goal 1, thinking we could have a check for wals in `MergeStats.verifyMergeConsistency()` that returns false.  This could be done in 2.1 and 3.1.  It may be redundant in 2.1, but would not hurt.   If we have that, then when a tablet has wals and the merge state is WAITING_FOR_OFFLINE then the tablet goal state should be HOSTED so it can be recovered.

For goal 2, if we are not doing that validation in the current code then I think it would be good to add it in 2.1 and 3.1

I tried to think of ways to test this scenario in an IT and could not think of a good way to do it.

**Comment by cshannon** [comment]
@keith-turner - I think your proposed changes make sense. I think adding checks to the verifyMergeConsistency() make the most sense as it can be shared in 2.1 and 3.1 and the purpose of the method is a verification before starting so it's the logical spot. As you said, in 2.1 it may be more redundant since the WALs are checked before but wouldn't hurt. It's definitely needed in 3.1 since those checks were removed when I removed chopped compactions. For testing, I agree, it may not be easy to test with an IT, we may just need some unit testing. I can start working on a PR for the changes.

**Comment by cshannon** [comment]
This is done in 2.1.3 and merged into main for 3.1.0. I merged an empty commit forward into elasticity as the changes are going to be completely different so it's best to create a new PR which I will work on. After elasticity is done we can close this issue.
